src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,906 @@
+         */
+        if (null == context.timestamp || 0 >= context.timestamp.length) {
+            context.token.put(context.targetInfo, 0, context.targetInfo.length);
+        } else {
+            ByteBuffer newTargetInfo = ByteBuffer.allocate(context.targetInfo.length + NTLM_MIC_AVP_LENGTH)
+                    .order(ByteOrder.LITTLE_ENDIAN);
+
+            // copy targetInfo up to NTLM_AVID_MSVAVEOL
+            newTargetInfo.put(context.targetInfo, 0, context.targetInfo.length - 4);
+
+            // MIC flag
+            newTargetInfo.putShort(NTLM_AVID_MSVAVFLAGS);
+            newTargetInfo.putShort((short) 4);
+            newTargetInfo.putInt((int) NTLM_AVID_VALUE_MIC);
+
+            // EOL
+            newTargetInfo.putShort(NTLM_AVID_MSVAVEOL);
+            newTargetInfo.putShort((short) 0);
+
+            context.token.put(newTargetInfo.array(), 0, newTargetInfo.array().length);
+        }
+
+        return context.token.array();
+    }
+
+    /**
+     * Gets the HMAC MD5 hash
+     *
+     * @see <a href="https://www.ietf.org/rfc/rfc2104.txt">https://www.ietf.org/rfc/rfc2104.txt</a>
+     * 
+     * @param key
+     * @param data
+     * @return HMAC MD5 hash
+     * @throws InvalidKeyException
+     */
+    private byte[] hmacMD5(final byte[] key, final byte[] data) throws InvalidKeyException {
+        SecretKeySpec keySpec = new SecretKeySpec(key, "HmacMD5");
+        context.mac.init(keySpec);
+        return context.mac.doFinal(data);
+    }
+
+    /**
+     * Gets the MD4 hash of input string
+     *
+     * @param str
+     *        input string
+     * @return MD4 hash
+     */
+    private byte[] MD4(final byte[] str) {
+        MD4 md = new MD4();
+        md.reset();
+        md.update(str);
+        return md.digest();
+    }
+
+    /**
+     * Gets the unicode of string
+     *
+     * @param str
+     *        string to convert to unicode
+     * @return unicode of string
+     */
+    private byte[] unicode(final String str) {
+        return str.getBytes(java.nio.charset.StandardCharsets.UTF_16LE);
+    }
+
+    /**
+     * Concatenates 2 byte arrays
+     *
+     * @param arr1
+     * @param arr2
+     * @return concatenated array
+     */
+    private byte[] concat(final byte[] arr1, final byte[] arr2) {
+        if (null == arr1 || null == arr2) {
+            return null;
+        }
+
+        byte[] temp = new byte[arr1.length + arr2.length];
+        System.arraycopy(arr1, 0, temp, 0, arr1.length);
+        System.arraycopy(arr2, 0, temp, arr1.length, arr2.length);
+        return temp;
+    }
+
+    /**
+     * Generates a key from password to get NTLMv2 hash Section 3.3.2 NTLM v2 Authentication
+     *
+     * @return NT response key
+     * @throws InvalidKeyException
+     */
+    private byte[] ntowfv2() throws InvalidKeyException {
+
+        return hmacMD5(context.passwordHash, (null != context.userName) ? unicode(context.userName + context.domainName)

```

Comment:
Can username be `null` here?

Reply:
you're right it can't because it should be set to default value in SQLServerConnection. But, good style to null check anyway?